### PR TITLE
Bitwise Addition

### PR DESCRIPTION
**Pull request by Beakerboy**

### CODE REVIEW

File: tests/Functions/BitwiseTest.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public function dataProviderForBitwiseAdd()
+    {
+        return [
+            [
+                1, 1, [false, 2]
+            ],
+            [
+                1, -1, [true, 0]
```
Comment (markrogoyski) [inline tests/Functions/BitwiseTest.php]: Why would overflow be true here?

File: tests/Functions/BitwiseTest.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            ],
+            [
+                \PHP_INT_MAX, 1, [false, \PHP_INT_MIN]
```
Comment (markrogoyski) [inline tests/Functions/BitwiseTest.php]: Why is overflow false here?

File: tests/Functions/BitwiseTest.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            ],
+            [
+                -1, -1, [true, -2]
```
Comment (markrogoyski) [inline tests/Functions/BitwiseTest.php]: Why is overflow true here?

File: src/Functions/Bitwise.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+     *
+     * @return array
+     *    'overflow' is true if the result is larger than the bits in an int
+     *    'value' is the result of the addition ignoring any overflow.
+     */
+    public static function bitwiseAdd(int $a, int $b): array
+    {
+        if (is_int($a + $b)) {
+            $sum = $a + $b;
+            $overflow = (($a < 0 || $b < 0) && $sum >= 0) || ($a < 0 && $b < 0);
```
Comment (markrogoyski) [inline src/Functions/Bitwise.php]: I'm not sure I understand the definition of overflow here. If the result of $a + $b is an int, then by definition, there should be no integer overflow, correct?

Comment (Beakerboy) [inline src/Functions/Bitwise.php]: This is bit overflow, not integer overflow. The comment block for the method gives some bitwise examples. In the 4 bit world, 0b1111 + 0b0001 will overflow...however, in decimal, this in -1 + 1...which would not overflow. The bits overflow while the ints do not.

Comment (markrogoyski) [inline src/Functions/Bitwise.php]: OK. Thanks for the explanation. I get it now. I didn't see any binary numbers or bit shifting so I was a bit (pun intended) confused. It's interesting that you can figure out the overflow without converting to binary/2's complement and doing the operation just with this conditional statement.

Comment (Beakerboy) [inline src/Functions/Bitwise.php]: The textual description of this statement is...if you add a negative and positive number together, and the result is positive, the MSB on the negative number had to have been shifted off. Anytime two negative numbers are added, the MSB on both is a '1', and 1+1 = 10, so there is a bit overflow. It took me a while to make this statement it as compact as I could.

Comment (markrogoyski) [inline src/Functions/Bitwise.php]: Thanks again for the explanation. Do you mind adding a little more documentation and/or comments about the three cases to explain what is going on and why is doing what it is doing. Thanks.

**Comment by markrogoyski** [comment]
Thanks for the pull request, and thanks for adding the explanatory documentation. I refactored a little bit to make the function a little cleaner, and changed the method name to just add as bitwiseAdd was redundant with the class name since it is a static method.

**Pull request merged**